digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/csvDataExporter/AssetReportCsvExporter.scala

```diff
@@ -0,0 +1,264 @@
+      if(withTransaction) {
+        withDynSession {
+          val userResults = userProvider.getUsers()
+          val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+          (userResults, municResults)
+        }
+      }
+      else{
+        val userResults = userProvider.getUsers()
+        val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+        (userResults, municResults)
+      }
+    }
+
+
+    val (users, municipalitiesInfo) = getUserAndMunicipalitiesInfo()
+
+    val operatorUsers = users.filter(_.isOperator())
+    val operatorUsersUsername = operatorUsers.map(_.username)
+
+    municipalitiesList.flatMap{ municipality =>
+
+      val municipalityUsers = users.filter( _.isAuthorizedToWrite(municipality) )
+                                    .filterNot( user => operatorUsersUsername.contains(user.username) )
+                                    .map( _.username )
+
+      val linearAssets = extractData(municipality)
+
+      if ( linearAssets != null && linearAssets.nonEmpty) {
+        val linearAssetsGroupedBy = linearAssets.groupBy(_.assetType)
+
+        assetTypesList.flatMap { assetType =>
+          if (linearAssetsGroupedBy.contains(assetType)) {
+            val currentAssetType = linearAssetsGroupedBy(assetType)
+
+            val (operatorUser, operatorDate) = extractUserAndDateTime(currentAssetType, operatorUsersUsername)
+            val (municipalityUser, municipalityDate) = extractUserAndDateTime(currentAssetType, municipalityUsers)
+
+            val firstElem = currentAssetType.head
+
+            Some( ExportAssetReport(municipalitiesInfo.filter(_.id == municipality).head.name, firstElem.assetType, currentAssetType.size,
+                              firstElem.assetNameFI, firstElem.assetGeometryType, operatorUser, operatorDate, municipalityUser, municipalityDate) )
+          }
+          else
+            None
+        }
+
+      }
+      else
+        None
+    }
+  }
+
+
+  def extractDateToString(dateTime: Option[DateTime]): String = {
+    dateTime match {
+      case Some(date) =>
+        DateParser.dateToString(date, DateParser.DatePropertyFormat)
+      case _ => ""
+    }
+  }
+
+
+  def extractRecords( exportAssetReports: List[ExportAssetReport], withTransaction: Boolean = true): String = {
+
+    def getTotalNewLawTrafficSigns(municipalityName: String): Int = {
+      if(withTransaction) {
+        withDynSession {
+          val municipalityInfo = municipalityDao.getMunicipalityIdByName(municipalityName)
+          assetReporterDAO.getTotalTrafficSignNewLaw(municipalityInfo.head.id)
+        }
+      }
+      else{
+        val municipalityInfo = municipalityDao.getMunicipalityIdByName(municipalityName)
+        assetReporterDAO.getTotalTrafficSignNewLaw(municipalityInfo.head.id)
+      }
+    }
+
+
+    val output = if (exportAssetReports.isEmpty) {
```

Comment:
I don't think this if isEmpty is necessary. The .map will return a empty list in it's own

Reply:
If I can easly check if the list is empty or not why should leave that to be validate in .map ?

Reply:
as requested